lib/optimizer/src/Transformer/AmpBoilerplate.php

```diff
@@ -70,23 +70,26 @@ public function transform(Document $document, ErrorCollection $errors)
     /**
      * Remove all <style> and <noscript> tags except for the <style amp-custom> tag.
      *
-     * @param Document $document Document to remove the tags from.
+     * @param Document $document    Document to remove the tags from.
+     * @param string   $boilerplate Boilerplate attribute name.
      */
-    private function removeStyleAndNoscriptTags(Document $document)
+    private function removeStyleAndNoscriptTags(Document $document, $boilerplate)
     {
         $headNode = $document->head->firstChild;
-        $devMode = DevMode::isActiveForDocument($document);
         while ($headNode) {
             $nextSibling = $headNode->nextSibling;
-            if ($headNode instanceof DOMElement && (! $devMode || ! DevMode::hasExemptionForNode($headNode))) {
+            if ($headNode instanceof DOMElement) {
                 switch ($headNode->tagName) {
                     case Tag::STYLE:
-                        if (! $headNode->hasAttribute(Attribute::AMP_CUSTOM)) {
+                        if ($headNode->hasAttribute($boilerplate)) {
                             $document->head->removeChild($headNode);
                         }
                         break;
                     case Tag::NOSCRIPT:
-                        $document->head->removeChild($headNode);
+                        $style = $headNode->getElementsByTagName('style')->item(0);
+                        if ($style instanceof DOMElement && $style->hasAttribute($boilerplate)) {
+                            $document->head->removeChild($headNode);
+                        }
```

Comment:
@schlessera this fixes something I noticed before when first reviewing the Optimizer code. I thought at some point the logic of removing all styles _other_ than the `amp-custom` could cause problems, and it seems my suspicion  was correct. So now this logic removes _exclusveily_ the AMP boilerplate styles only, leaving other styles intact.

Reply:
I wasn't too happy with this logic either, but it is what the canonical Go optimizer package uses as well:
https://github.com/ampproject/amppackager/blob/releases/transformer/transformers/ampboilerplate.go#L24-L37

Maybe an issue and/or PR for the Go package would make sense as well?

Reply:
Yes. But otherwise, you see the changes in this PR as correct and approved?

Reply:
(Our needs are a bit different in that we often need to serve invalid AMP, whereas the Go version probably only ever needs to serve valid AMP. So that's perhaps why it is fine for Go use the logic it does now, but otherwise it seems more robust to do what this PR does now.)

Reply:
One problem that I see is that if there's a mismatch between the type of AMP that the `<html>` tag states, and the boilerplate that is already included, the wrong one will stay in place.

So, if the document states it is AMP4EMAIL, but the AMP4ADS boilerplate was added (for whatever reason), the optimizer will only try to remove the AMP4EMAIL boilerplate (and fail), and then re-insert it, with the end result being that two boilerplates are in the document.

Reply:
I was originally looping over all the possible boilerplates but then paired it down to just the identified one. You think it's best to check all?

Reply:
Yes, I think that'd be safer. We should always keep in mind that the Optimizer is not at all coupled to the plugin, so we cannot make any assumptions about the actual HTML that gets served into the optimizer. For all we know, it could be a buggy Drupal extension or a Symfony middleware that hands it over to the Optimizer. So, while it's important to keep the needs of the plugin in mind, we also need to keep the Optimizer in a state where it will just work regardless of input.

Reply:
How do you think of 676205438?

Reply:
Approach looks good, just not happy with the naming (see code review comments).